File: ts_scrapers/scrapers/halkbank/paraf.ts

import puppeteer from 'puppeteer-extra';
import StealthPlugin from 'puppeteer-extra-plugin-stealth';
import { createClient } from '@supabase/supabase-js';
import * as dotenv from 'dotenv';
import { parseWithGemini } from '../../services/geminiParser';
import { generateSectorSlug, generateCampaignSlug } from '../../utils/slugify';
import { syncEarningAndDiscount } from '../../utils/dataFixer';
import { normalizeBankName, normalizeCardName } from '../../utils/bankMapper';
import { optimizeCampaigns } from '../../utils/campaignOptimizer';
import { lookupIDs } from '../../utils/idMapper';
import { assignBadge } from '../../services/badgeAssigner';
import { markGenericBrand } from '../../utils/genericDetector';

puppeteer.use(StealthPlugin());

dotenv.config();

const supabase = createClient(
    process.env.SUPABASE_URL!,
    process.env.SUPABASE_ANON_KEY!
);

const BASE_URL = 'https://www.paraf.com.tr';
const CAMPAIGNS_URL = 'https://www.paraf.com.tr/tr/kampanyalar.html';

const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

async function runParafScraper() {
    console.log('\n💳 Paraf (Halkbank)');
    const normalizedBank = await normalizeBankName('Halkbank');
    const normalizedCard = await normalizeCardName(normalizedBank, 'Paraf');
    console.log(`   Bank: ${normalizedBank}, Card: ${normalizedCard}`);

    const args = process.argv.slice(2);
    const limitArg = args.find(arg => arg.startsWith('--limit='));
    const limit = limitArg ? parseInt(limitArg.split('=')[1]) : 999;
    const isAIEnabled = args.includes('--ai');

    const browser = await puppeteer.launch({
        headless: true,
        args: [
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage',
            '--disable-blink-features=AutomationControlled'
        ]
    });

    try {
        const page = await browser.newPage();
        await page.setViewport({ width: 1920, height: 1080 });
        await page.setUserAgent('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');

        // 1. Load campaign list and click "Load More"
        console.log(`   📄 Loading ${CAMPAIGNS_URL}...`);
        await page.goto(CAMPAIGNS_URL, { waitUntil: 'networkidle2', timeout: 60000 });

        // Wait for campaign list
        try {
            await page.waitForSelector('.cmp-list--campaigns', { timeout: 20000 });
        } catch (e) {
            console.log('   ⚠️ Warning: .cmp-list--campaigns not found within timeout. Page might have loaded differently.');
        }

        // Click "Load More" button multiple times
        let clickCount = 0;
        const maxClicks = 30;

        while (clickCount < maxClicks) {
            try {
                // Selector update: some buttons might be nested or have slightly different classes
                // Verified current: .button--more-campaign a
                const loadMoreBtn = await page.$('.button--more-campaign a, .button--more-campaign .cmp-button');
                if (!loadMoreBtn) break;

                const isVisible = await loadMoreBtn.evaluate(el => {
                    const style = window.getComputedStyle(el);
                    return style && style.display !== 'none' && style.visibility !== 'hidden' && el.offsetHeight > 0;
                });

                if (!isVisible) break;

                await loadMoreBtn.scrollIntoView();
                await sleep(1000);
                await loadMoreBtn.click();
                await sleep(3000); // Increased wait for dynamic content
                clickCount++;
                console.log(`   -> Clicked 'Load More' (${clickCount})`);
            } catch (e) {
                console.log(`   -> Stopping 'Load More' after ${clickCount} clicks due to error or missing button.`);
                break;
            }
        }

        // Extract campaign links
        const campaignLinks = await page.evaluate((baseUrl) => {
            const links: string[] = [];
            // Primary selector from reference
            const anchors = document.querySelectorAll('.cmp-list--campaigns .cmp-teaser__title a, .cmp-teaser__title a');
            anchors.forEach(a => {
                const href = (a as HTMLAnchorElement).getAttribute('href');
                if (href && (href.includes('/kampanyalar/') || href.includes('/content/parafcard/'))) {
                    if (href.endsWith('kampanyalar.html')) return;
                    const fullUrl = href.startsWith('http') ? href : baseUrl + (href.startsWith('/') ? '' : '/') + href;
                    if (!links.includes(fullUrl)) {
                        links.push(fullUrl);
                    }
                }
            });
            return links;
        }, BASE_URL);

        console.log(`\n   🎉 Found ${campaignLinks.length} campaigns.`);

        if (campaignLinks.length === 0) {
            console.log('   ❌ No campaign links found. Check selectors!');
            return;
        }

        // Apply limit
        const limitedLinks = limit ? campaignLinks.slice(0, limit) : campaignLinks;
        console.log(`   Processing first ${limitedLinks.length}...`);

        // 2. Optimize
        const { urlsToProcess } = await optimizeCampaigns(limitedLinks, normalizedCard);

        // 3. Process each campaign
        for (const fullUrl of urlsToProcess) {
            console.log(`\n   🔍 Processing: ${fullUrl}`);

            try {
                await page.goto(fullUrl, { waitUntil: 'networkidle2', timeout: 30000 });
                await sleep(1000);

                // Wait for title
                try {
                    await page.waitForSelector('h1', { timeout: 10000 });
                } catch { }

                const html = await page.content();

                // Extract fallback data
                const fallbackData = await page.evaluate((baseUrl) => {
                    const titleEl = document.querySelector('.master-banner__content h1') || document.querySelector('h1');
                    const title = titleEl ? titleEl.textContent?.trim() : 'Başlıksız Kampanya';

                    // Extract image
                    let image: string | null = null;
                    const bannerDiv = document.querySelector('.master-banner__image') as HTMLElement;
                    if (bannerDiv && bannerDiv.style.backgroundImage) {
                        const match = bannerDiv.style.backgroundImage.match(/url\(['"]?(.*?)['"]?\)/);
                        if (match && !match[1].includes('logo.svg')) {
                            const imgPath = match[1];
                            image = imgPath.startsWith('http') ? imgPath : baseUrl + (imgPath.startsWith('/') ? '' : '/') + imgPath;
                        }
                    }

                    if (!image) {
                        const imgs = Array.from(document.querySelectorAll('img'));
                        for (const img of imgs) {
                            const src = img.getAttribute('src') || img.getAttribute('data-src');
                            if (src && !src.includes('logo') && !src.includes('icon') && !src.includes('.svg')) {
                                if (src.includes('/content/') || src.includes('kampanyalar')) {
                                    image = src.startsWith('http') ? src : baseUrl + (src.startsWith('/') ? '' : '/') + src;
                                    break;
                                }
                            }
                        }
                    }

                    if (!image) {
                        image = 'https://www.paraf.com.tr/content/dam/parafcard/paraf-logos/paraf-logo-yeni.png';
                    }

                    return { title, image };
                }, BASE_URL);

                let campaignData: any;

                if (isAIEnabled) {
                    campaignData = await parseWithGemini(html, fullUrl, normalizedBank, normalizedCard);

                    // Fallback title if AI missed it
                    if (!campaignData.title && fallbackData.title) {
                        campaignData.title = fallbackData.title;
                    }

                    // Fallback image if AI missed it
                    if (!campaignData.image && fallbackData.image) {
                        campaignData.image = fallbackData.image;
                        console.log('      🔧 Used fallback image');
                    }

                    // Paraf-specific: Default participation method if missing
                    if (!campaignData.participation_method) {
                        if (html.includes('Paraf Mobil') || html.includes('Halkbank Mobil') || html.includes('HEMEN KATIL')) {
                            campaignData.participation_method = "Paraf Mobil veya Halkbank Mobil uygulamasından 'Hemen Katıl' butonuna tıklayın";
                        } else if (html.includes('3404')) {
                            const smsMatch = html.match(/([A-Z0-9]{3,})\s*yazıp\s*3404/i);
                            if (smsMatch) {
                                campaignData.participation_method = `SMS (${smsMatch[1].toUpperCase()} -> 3404)`;
                            }
                        }
                    }
                } else {
                    // No AI mode
                    campaignData = {
                        title: fallbackData.title,
                        description: fallbackData.title,
                        image: fallbackData.image,
                        category: 'Diğer',
                        sector_slug: 'genel',
                        card_name: normalizedCard,
                        bank: normalizedBank,
                        url: fullUrl,
                        reference_url: fullUrl,
                        is_active: true,
                        tags: []
                    };
                }

                if (campaignData) {
                    // Ensure critical fields
                    campaignData.card_name = normalizedCard;
                    campaignData.bank = normalizedBank;
                    campaignData.url = fullUrl;
                    campaignData.reference_url = fullUrl;
                    if (!campaignData.sector_slug) campaignData.sector_slug = 'genel';

                    syncEarningAndDiscount(campaignData);
                    campaignData.publish_status = 'processing';
                    campaignData.publish_updated_at = new Date().toISOString();

                    // Set default min_spend
                    if (campaignData.min_spend === undefined || campaignData.min_spend === null) {
                        campaignData.min_spend = 0;
                    }

                    // Lookup IDs
                    const idsResult = await lookupIDs(
                        campaignData.bank,
                        campaignData.card_name,
                        campaignData.brand,
                        campaignData.sector_slug,
                        campaignData.category
                    );
                    Object.assign(campaignData, idsResult);

                    // Assign badge
                    const badgeResult = assignBadge(campaignData);
                    campaignData.badge_text = badgeResult.text;
                    campaignData.badge_color = badgeResult.color;

                    // Mark generic brand
                    markGenericBrand(campaignData);

                    campaignData.tags = campaignData.tags || [];


                    // Upsert to database

                    // ID-BASED SLUG SYSTEM
                    const { data: existing } = await supabase
                        .from('campaigns')
                        .select('id')
                        .eq('reference_url', fullUrl)
                        .single();

                    if (existing) {
                        const finalSlug = generateCampaignSlug(campaignData.title, existing.id);
                        const { error } = await supabase
                            .from('campaigns')
                            .update({ ...campaignData, slug: finalSlug })
                            .eq('id', existing.id);
                        if (error) {
                            console.error(`      ❌ Update Error: ${error.message}`);
                        } else {
                            console.log(`      ✅ Updated: ${campaignData.title} (${finalSlug})`);
                        }
                    } else {
                        const { data: inserted, error: insertError } = await supabase
                            .from('campaigns')
                            .insert(campaignData)
                            .select('id')
                            .single();
                        if (insertError) {
                            console.error(`      ❌ Insert Error: ${insertError.message}`);
                        } else if (inserted) {
                            const finalSlug = generateCampaignSlug(campaignData.title, inserted.id);
                            await supabase
                                .from('campaigns')
                                .update({ slug: finalSlug })
                                .eq('id', inserted.id);
                            console.log(`      ✅ Inserted: ${campaignData.title} (${finalSlug})`);
                        }
                    }
                }

            } catch (error: any) {
                console.error(`      ❌ Error processing: ${error.message}`);
                continue;
            }
        }

        console.log('\n✅ Scraper finished.');

    } catch (error: any) {
        console.error(`❌ Fatal error: ${error.message}`);
        process.exit(1);
    } finally {
        await browser.close();
    }
}

runParafScraper();

File: ts_scrapers/scrapers/halkbank/paraf.py

import time
import json
import random
import os
import sys
import ssl
import undetected_chromedriver as uc

# MacOS SSL Fix
ssl._create_default_https_context = ssl._create_unverified_context

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from bs4 import BeautifulSoup

# --- CONFIGURATION ---
BASE_URL = "https://www.paraf.com.tr"
START_URL = "https://www.paraf.com.tr/tr/kampanyalar.html"
OUTPUT_FILE = "paraf_kampanyalar_raw.json"
# Parse limit from args
CAMPAIGN_LIMIT = 1000
if len(sys.argv) > 1 and "--limit" in sys.argv:
    try:
        idx = sys.argv.index("--limit")
        CAMPAIGN_LIMIT = int(sys.argv[idx + 1])
    except: pass
print(f"   🎯 Limit: {CAMPAIGN_LIMIT}")

def get_random_user_agent():
    user_agents = [
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Safari/605.1.15"
    ]
    return random.choice(user_agents)

def setup_driver():
    options = uc.ChromeOptions()
    options.add_argument(f"user-agent={get_random_user_agent()}")
    options.add_argument("--no-sandbox")
    options.add_argument("--disable-dev-shm-usage")
    options.add_argument("--window-size=1920,1080")
    # options.add_argument("--headless=new") # Debug için kapalı tutuyoruz, gerekirse açılır

    driver = uc.Chrome(options=options)
    return driver

def scroll_and_click_more(driver):
    print("   🔄 Scroll ve 'Daha Fazla' butonu kontrol ediliyor...")
    
    # Initial load wait
    time.sleep(5) 
    
    click_count = 0
    max_clicks = 30
    if CAMPAIGN_LIMIT < 20: max_clicks = 1  # Optimize for testing
    
    while click_count < max_clicks:
        try:
            # Scroll to bottom to trigger events
            driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
            time.sleep(2)
            
            # Find visible "Daha Fazla" buttons
            buttons = driver.find_elements(By.CSS_SELECTOR, ".button--more-campaign a, a.btn-more, button")
            clicked = False
            
            for btn in buttons:
                if btn.is_displayed() and "DAHA FAZLA" in btn.text.strip().upper():
                    print(f"   👇 'Daha Fazla'ya tıklanıyor ({click_count+1})...")
                    driver.execute_script("arguments[0].click();", btn)
                    time.sleep(3) # Wait for content
                    clicked = True
                    click_count += 1
                    break
            
            if not clicked:
                print("   ✅ 'Daha Fazla' butonu kalmadı.")
                break
                
        except Exception as e:
            print(f"   ⚠️ Scroll Hatası: {e}")
            break

def scrape_list_page(driver):
    print(f"   🌐 Liste taranıyor: {START_URL}")
    driver.get(START_URL)
    scroll_and_click_more(driver)
    
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    links = []
    
    # Paraf Reference Selector: .cmp-list--campaigns .cmp-teaser__title a
    items = soup.select('.cmp-list--campaigns .cmp-teaser__title a')
    
    # 2. Backup Selektörler
    if not items:
        print("   ⚠️ Primary selector failed. Trying backup...")
        items = soup.select('a[href*="/kampanyalar/"]')
        
    print(f"   🔍 Bulunan link sayısı: {len(items)}")
    
    for item in items:
        href = item.get('href')
        if href and "/kampanyalar/" in href and not href.endswith("kampanyalar.html"):
            full_url = href if href.startswith("http") else BASE_URL + href
            if full_url not in links:
                links.append(full_url)
                
    return links[:CAMPAIGN_LIMIT]

def robust_get(driver, url, max_retries=3):
    for attempt in range(max_retries):
        try:
            driver.get(url)
            # Wait for meaningful content
            WebDriverWait(driver, 20).until(
                EC.presence_of_element_located((By.CSS_SELECTOR, "h1"))
            )
            return True
        except Exception as e:
            print(f"      ⚠️ Bağlantı hatası ({attempt+1}/{max_retries}): {str(e)[:50]}...")
            try:
                driver.delete_all_cookies()
            except: pass
            time.sleep(random.uniform(10, 20))
    return False

def scrape_detail(driver, url):
    if not robust_get(driver, url):
        print("      ❌ Sayfa yüklenemedi, atlanıyor.")
        return None

    soup = BeautifulSoup(driver.page_source, 'html.parser')
    
    # 1. Title
    title_el = soup.select_one('.master-banner__content h1') or soup.select_one('h1')
    title = title_el.text.strip() if title_el else "Başlıksız Kampanya"
    
    # 2. Image
    image = ""
    # Try style attribute first (common in Paraf)
    img_div = soup.select_one('.master-banner__image')
    if img_div and 'style' in img_div.attrs:
        import re
        m = re.search(r'url\([\'"]?(.*?)[\'"]?\)', img_div['style'])
        if m: image = m.group(1)
        
    # Backup image
    if not image:
        # Try finding valid content images
        all_imgs = soup.find_all('img')
        for img in all_imgs:
            src = img.get('src') or img.get('data-src') or ""
            if src:
                src_lower = src.lower()
                # Filter out junk
                if "logo" in src_lower or "icon" in src_lower or ".svg" in src_lower:
                    continue
                if "placeholder" in src_lower or "spacer" in src_lower:
                    continue
                    
                # Must be a content image
                if "/content/" in src or "campaigns" in src_lower:
                    image = src
                    break
        
    if image and not image.startswith('http'):
        image = BASE_URL + image
        
    # Final Fallback (if still empty)
    if not image or "logo" in image.lower():
         image = "https://www.paraf.com.tr/content/dam/parafcard/paraf-logos/paraf-logo-yeni.png"

    # 3. Description (Short) & Detail HTML (Long)
    # Paraf content is usually in .cmp-text
    content_div = soup.select_one('.text--use-ulol .cmp-text') or soup.select_one('.cmp-text')
    
    description = ""
    detail_html = ""
    
    if content_div:
        # P tags for short description
        ps = content_div.find_all('p')
        description = " ".join([p.text.strip() for p in ps[:2]]) # First 2 paragraphs as summary
        detail_html = str(content_div) # Keep full HTML for AI
    else:
        # Fallback to body text if no container found
        description = title
        detail_html = str(soup.body)

    return {
        "url": url,
        "title": title,
        "description": description,
        "detail_html": detail_html, # Critical for AI
        "image": image,
        "bank": "Halkbank",
        "card": "Paraf"
    }

def main():
    print("🚀 Paraf Python Scraper Başlatılıyor (Hybrid Mode)...")
    driver = setup_driver()
    
    try:
        links = scrape_list_page(driver)
        print(f"   🎯 Toplam {len(links)} kampanya işlenecek.")
        
        results = []
        for i, link in enumerate(links):
            print(f"   [{i+1}/{len(links)}] İşleniyor: {link}")
            data = scrape_detail(driver, link)
            if data:
                results.append(data)
                # Save continually
                with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
                    json.dump(results, f, ensure_ascii=False, indent=4)
            time.sleep(random.uniform(2, 5)) # Polite delay
            
        # Final Save
        with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
            json.dump(results, f, ensure_ascii=False, indent=4)
            
        print(f"\n✅ İşlem Tamamlandı! {len(results)} kampanya kaydedildi: {OUTPUT_FILE}")
        
    except Exception as e:
        print(f"\n❌ Kritik Hata: {e}")
    finally:
        driver.quit()

if __name__ == "__main__":
    main()
